packages/liveblocks-react-ui/src/components/Composer.tsx

```diff
@@ -555,9 +570,53 @@ export const Composer = forwardRef(
     forwardedRef: ForwardedRef<HTMLFormElement>
   ) => {
     const client = useClient();
-    const createThread = useCreateThread();
-    const createComment = useCreateComment();
-    const editComment = useEditComment();
+    const thread = useContext(ThreadDataContext);
+    const comment = useContext(CommentDataContext);
+    const room = useRoomOrNull();
+
+    function getRoomId(): string {
+      // Both threadId and commentId are provided (editing a comment)
+      if (threadId !== undefined && commentId !== undefined) {
+        if (comment && comment.id === commentId) {
+          return comment.roomId;
+        } else if (thread && thread.id === threadId) {
+          return thread.roomId;
+        } else if (room) {
+          return room.id;
+        } else {
+          throw new Error(
+            "Composer is provided both threadId and commentId but is not a descendant of Comment, Thread, or RoomProvider component"
+          );
+        }
+      }
+      // Only threadId is provided (replying to a thread)
+      else if (threadId !== undefined) {
+        if (thread && thread.id === threadId) {
+          return thread.roomId;
+        } else if (room) {
+          return room.id;
+        } else {
+          throw new Error(
+            "Composer is provided a threadId but is not a descendant of Thread or RoomProvider component"
+          );
+        }
+      }
+      // Default to room context (creating a new thread)
+      else if (room) {
+        return room.id;
+      }
+
+      // No valid context found
+      throw new Error(
+        "Composer must be a descendant of RoomProvider component"
+      );
+    }
+
+    const roomId = getRoomId();
+
+    const createThread = useCreateRoomThread(roomId);
+    const createComment = useCreateRoomComment(roomId);
+    const editComment = useEditRoomComment(roomId);
```

Comment:
Notice these private hooks that don't internally rely on `useRoom()` hook. This made the refactoring easier to reason about. The only thing we had to worry about was figuring out where the `roomId` should come from, and that is described here: https://github.com/liveblocks/liveblocks/pull/2028/files#r1843300206